### ISSUE TITLE
Car accidents (separate mod?)

### ISSUE THREAD

**Issue by aubergine10** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#61]
On Steam, you mentioned this should be separate mod, and I agree.

However, IMO, it will need tight integration with TM:PE so just jotting down some thoughts here....

Choosing accident sites:

* Reckless drivers could be the culprit
* Dangerous junctions (see: #4 - high conflict score) = increase risk
* Fast roads = increase risk
* Poor road condition = increase risk
* Bad weather = increase risk
* Sharp bends = increas risk
* Heavy congestion or traffic jam = increase risk
* Roadworks (potential future feature?) = increase risk
* Industrial areas = increase risk
* etc...

Making the accident happen:

* Some way to make vehicle 'crash'
    * Procedural Objects mod could provide inspiration
* Visual effect: fire / smoke / steam (depending on severity of crash)
    * Spawn points (props) for these exist in vanilla game
* Mark affected lane(s) or segment(s) as "blocked" (like a building fire does)
    * This causes a traffic jam, and other vehicles try and route around it
* Call emergency services
    * What if they can't reach crash site due to traffic jam?
        * Call helicopter?
        * Route via other side of road?
        * Temporary lane routing could help traffic divert around crash?
    * Fire service if car is badly damaged or on fire
    * Ambulance if injuries
    * Hearse if deaths
    * Disaster recovery if it's major carnage
* Afterwards:
    * Clear visual effects
    * Remove 'blocked' flag from lane(s) or segment(s)

Additional visualisation:

* There are assets for overhead gantries, LED information displays and so on
* Would be nice if TMPE had tool to add those and make them actually work (by swapping props to display different messages)
    * Example: TMPE will know where congestion is, and so on
* Accidents could cause the LED displays to indicate lane closures, delays, etc.

Things to consider:

* TMPE tracks road congestion - crashes would potentially distort the stats for days?

**Comment by krzychu124** [comment]
Many thanks for your thoughts. I would like to see that in game in the future. 
I agree with you that would need tight integration but it's still possible to call TM:PE functions from outside from any other mod if we set that functions to public scope. In this case any data we are collecting about road usage and actual traffic lights state would be easily accessible.

In my opinion some of your ideas could greatly help with development of this epic feature (#14 , #26 ).
In addition i've been thinking about making cars to pull over if emergency vehicle is coming from behind to decrease time needed to reach their destination.

As far as I know we could also use built-in Transfer Manager to create offers for emergency services. That would solve problems with managing paths to accident location.

Traffic Manager 2.0 :smile:

**Comment by pcfantasy** [comment]
Nice idea，I also think about those ideas， let us make this happen in the future

**Comment by aubergine10** [comment]
The desire for accidents and breakdowns has been around since the game was launched, and people regularly mention it in respect to traffic mods, but also more generally in discussion forums, reddit, etc.

Vehicle breakdowns would be nice too. I see there's already a mod that makes vehicles require fuel; maybe they could break down if they run out before being able to refuel/recharge. It would likely need a new 'vehicle recovery' service (tow trucks, etc). There are some vehicle assets in the workshop that would suit such a service, they would need updating to have flashing lights when "on call", but building assets for the services would also be required. Towing vehicles would treat the broken down vehicle as a trailer to the towtruck. My assumption is that any vehicle can be towed by any other vehicle as far as the code is concerned.

Speaking of flashing "on call" lights, it's always bugged me how emergency vehicles turn their lights out as soon as they reach destination, rather than waiting until the emergency is fixed. This would be particularly problematic in the case of accident response; if there's a crash the emergency services keep their flashing lights on to warn drivers of the crash.

For making vehicles move out of the way for emergency vehicles, jfarias was experimenting with that back in T++ days. He mentioned having a basic implementation working, so there might be something lurking in a branch or tag of the [old T++ repo](https://github.com/joaofarias/csl-traffic). Vehicles could take advantage of parking lanes, bike lanes, bus lanes, etc. However, even with vehicles trying to move out of the way, there might still be some roads that don't facilitate such actions (like small 1u alley road from NeXT2).

At some point it might be worth doing some docs to help people integrate their mods with T++. Not API docs, maybe just pointers to where to look in the code and summaries of some useful functions with a few Do's and Don'ts? TMPE could possibly provide some helper functions for common things, like "is this junction dangerous" or "does this road have emergency lanes" and so on.

**Comment by pcfantasy** [comment]
@aubergine10

The mod you mentioned that makes vehicles require fuel is written by me. I tried to make vehicles break down if they run out before being able to refuel/recharge, but It is hard for me to implement items like you said such as  "other vehicles try and route around it"  or  "Visual effect: fire / smoke / steam (depending on severity of crash)"

**Comment by AdamDri** [comment]
@pcfantasy Visual effect can be seen in this mod https://steamcommunity.com/sharedfiles/filedetails/?id=780720853
I learned about it today

**Comment by aubergine10** [comment]
@AdamDri That mod adds the effect to all instances of a vehicle though? For car crashes I think it would be easier to just use the steam/smoke/fire/etc spawn points that come with the base game. Benefit is that they would automatically be included in save game (they are just props) and it would be easy to clear up after the crash (just remove those props within small radius of the accident).

**Comment by krzychu124** [comment]
There should be a way to spawn that effect on demand, I think

**Comment by aubergine10** [comment]
The smoke/steam/fire? Yes, just place the props, User can already spawn those props using "Find It" mod (search "steam" or "smoke" etc and you will find the props, you can just plop them on the map and they work instantly).

**Comment by pcfantasy** [comment]
Guys, Let me remind you something.

The key point for whatever car breakdown or car crash function is , we need to let cars change lanes not only in nodes.

In current game, cars can change lanes only inside nodes, so if a car breakdown, the other car behinds it can not change lanes to continue their trips.

**Comment by aubergine10** [comment]
Isn't that kind of what we want anyway? That breakdowns and accidents cause traffic jams?

As mentioned in OP, we could even mark road segments as blocked. (Note: I suspect that marking a segment as blocked will cause cars within that segment to do an in-road u-turn and leave the segment)

**Comment by pcfantasy** [comment]
@aubergine10

Ok, I understand your idea. I may try to write a mod first to test this.

**Comment by aubergine10** [comment]
Yes, it should definitely start as some small test mods just to try out different aspects of what would be required.

For example:

* a mod that spawns some smoke/steam/fire to test that those things work
* a mod that tests setting the 'blocked' flag on segments (I think it's on segments, but not sure)
    * what happens when the flag is set, what happens when it's removed?
    * note: emergency vehicles ignore blocked flag
* a mod that tests calling in emergency services to a specific location
    * Fire
    * Police
    * Ambulance
    * Disaster recovery
    * ... each one need to test: road vs. helicopter response (if possible to specify in the "Transfer Manager" bid things)
* etc...

Each of these little mods would be useful for other devs to look at as they can see in a small package how to achieve a certain thing. So sources should be in github.

**Comment by pcfantasy** [comment]
@aubergine10

So I quickly check the codes, blocked flag can add on segments, but if there is a car accident in 8-way road, we have to block the whole road?

That seems not real.

**Comment by aubergine10** [comment]
Yes, that is somewhat excessive. Ideally we only want to block on single side of road (and maybe reduce speed on other side of road as drivers will slow down to look at accident lol).

But if you can determine how the 'blocked' flag works, I mean, how it makes vehicles avoid that route and (I don't know if this happens, you'd need to test) if it makes cars already on the blocked segment turn around and find new route, then we could perhaps create a custom version of "blocked" that only affects one side of segment.

Note: "One side of a road" actually means "a set of **adjacent** road vehicle lanes going in same direction. (there are some roads that sort of split in multiple sections [like this](https://streetmix.net/-/723477))

**Comment by pcfantasy** [comment]
@aubergine10  
Do you know how do create a prop?

Is there any mod I can use as a reference?

**Comment by aubergine10** [comment]
Find It mod places props.

**Comment by pcfantasy** [comment]
@aubergine10 
No I mean how to use code to spawn a prop on the road.
There is a method to CreateProp, but I do not know how to get a PropInfo of a certain prop
public bool CreateProp(out ushort prop, ref Randomizer randomizer, PropInfo info, Vector3 position, float angle, bool single)

**Comment by aubergine10** [comment]
I don't know, sorry. Find It mod obviously has to get prop info to list props and allow user to place them on the map, so digging through source of that mod should help: https://github.com/SamsamTS/CS-FindIt

**Comment by Strdate** [comment]
You can disable pathfinding on the affected lanes and force recalculation when vehicles approach. What regards the visual part of this, when you check the Vehicle Effects mod you should find out how to spawn these visuals on selected vehicles only.

When you delete a road in vanilla game, the vehicles fly to other nearby road as was mentioned. Maybe in case of an accident if you removed lanes on one part of the road, vehicles would move to the other and it wouldn't even look like flying but be realistic enuf. But I don't know.

**Comment by aubergine10** [comment]
Car crashes could be implemented using custom simulation step in the CarAI, similar to how blown or flood vehicles are handled (branched off at the start of the main simulation step).

**Comment by aubergine10** [comment]
Related idea from @kvakvs : Vehicle breakdowns

> a friend of mine had an idea, not sure if he ever begins doing that. Make cars break sometimes, make car collisions and crashes. And not just general breakdown, but one of car parts can break producing different effects which you can observe (slow down, smoke, stop and block the lane, etc)

**Comment by r59andrey** [comment]
make it so that at crossroads without a traffic light, the transport was stupid (driving slowly).

**Comment by richaell** [comment]
It's been a while since this was mentioned. Any updates on ideas or new ways to implement these? @pcfantasy @krzychu124
